### PR TITLE
We need to convert libpod.conf files in user homedir for cgroupv2

### PR DESCRIPTION
**Pull request by rhatdan**
If a user upgrades to a machine that defaults to a cgroups V2 machine
and has a libpod.conf file in their homedir that defaults to OCI Runtime runc,
then we want to change it one time to crun.

runc as of this point does not work on cgroupV2 systems.  This patch will
eventually be removed but is needed until runc has support.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

### CODE REVIEW

**Comment by rhatdan** [comment]
Fixes: https://bugzilla.redhat.com/show_bug.cgi?id=1752040

**Comment by rhatdan** [comment]
@debarshiray PTAL
@mheon @giuseppe @vrothberg @TomSweeneyRedHat @baude @QiWang19 PTAL

**Comment by giuseppe** [comment]
do you think we should move the logic to `system migrate` or somehow rewrite the containers configuration to switch the runtime?  Existing containers won't work with cgroups v2 unless they are recreated

**Comment by mheon** [comment]
Agree with `system migrate`.

We might want to consider sneaking this in after RC1.

**Comment by rhatdan** [comment]
Can we trigger a system migrate without breaking every users containers?

File: libpod/runtime.go [hunk 1]
```diff
@@ -1451,3 +1466,36 @@ func (r *Runtime) ImageRuntime() *image.Runtime {
...
+	if !tmpConfig.CgroupCheck && rootless.IsRootless() {
+		cgroupsV2, err := cgroups.IsCgroup2UnifiedMode()
+		if err != nil {
+			return err
+		}
+		if cgroupsV2 {
+			path, err := exec.LookPath("crun")
+			if err != nil {
+				// Can't find crun path so do nothing
+				return nil
```
Comment (TomSweeneyRedHat) [inline libpod/runtime.go]: On the fence about having a debug statement in here too.  I'll let you make the call.

**Comment by TomSweeneyRedHat** [comment]
One question, lots of test unhappiness.

**Comment by debarshiray** [comment]
> Existing containers won't work with cgroups v2 unless they are recreated

Should I be worried about this? :)

Is a reboot followed by a `podman system migrate` enough to get past this? Or do the containers actually have to be recreated?

**Comment by TomSweeneyRedHat** [comment]
Fingers crossed, `podman system migrate` should handle it for you.  If @mheon says differently, go with his thinking.

**Comment by mheon** [comment]
I don't think that we expect existing containers to be broken after the config file change?

**Comment by mheon** [comment]
@rhatdan `system migrate` is explicitly for things like this - though config file changes have previously been out of its scope, I think it's the right place.

We should discuss rootless config files, though. This is not the only time we're going to hit this problem, I think - we don't have a good way of dealing with them within the package manager, and Podman has no idea if a user intentional set `runc` instead of `crun`...

**Comment by TomSweeneyRedHat** [comment]
Changes LGTM, but it looks like you're still fighting the good fight with the CI system

**Comment by rhatdan** [comment]
@giuseppe WDYT?

**Comment by mheon** [comment]
I'm starting to lean toward the suggestion that we not change the config file, but just ignore `runc` with CGroups V2 (but print a warning) for now - until it becomes safe.

**Comment by giuseppe** [comment]
> I'm starting to lean toward the suggestion that we not change the config file, but just ignore `runc` with CGroups V2 (but print a warning) for now - until it becomes safe.

I'd not ignore it, I think a warning should be enough.  I don't like much the idea of ignoring an user configuration when explicitly set

**Comment by rhatdan** [comment]
Then we blow up everyone in the universe and I would bet the exactl 0 people have actually said they want runc.  This is just because we don't merge the system defaults with user customization.  We need to change the processing of libpod to be grab the system default and then merge user changes.

**Comment by giuseppe** [comment]
> Then we blow up everyone in the universe and I would bet the exactl 0 people have actually said they want runc. This is just because we don't merge the system defaults with user customization. We need to change the processing of libpod to be grab the system default and then merge user changes.

good point.  Ok so let's get this in.  LGTM

**Comment by TomSweeneyRedHat** [comment]
Green test buttons on this one....

File: libpod/runtime.go [hunk 2]
```diff
@@ -254,6 +255,8 @@ type RuntimeConfig struct {
 	// SDNotify tells Libpod to allow containers to notify the host
 	// systemd of readiness using the SD_NOTIFY mechanism
 	SDNotify bool
+	// CgroupCheck verifies if the cgroup check for correct OCI runtime has been done.
+	CgroupCheck bool `toml:"cgroup_check,omitempty"`
```
Comment (mheon) [inline libpod/runtime.go]: Do we need the bool here? It doesn't seem to do much.

File: libpod/runtime.go (hunk 1 above)
Comment (mheon) [reply libpod/runtime.go]: Agree we should `Warnf` here about crun not existing - that's pretty serious

File: libpod/runtime.go (hunk 2 above)
Comment (mheon) [reply libpod/runtime.go]: Oh, nevermind, I see what you're doing here...

I don't really like it, but I don't see another easy way, and we can just remove it once we're done.

**Comment by mheon** [comment]
I'd second Tom's request for a bit more debugging, otherwise LGTM.

**Comment by rhatdan** [comment]
@mheon @TomSweeneyRedHat Added log message on failure to find crun.

**Comment by TomSweeneyRedHat** [comment]
LGTM and happy green test buttons

**Comment by rhatdan** [comment]
@mheon @vrothberg @giuseppe @baude PLEASE give this the /lgtm

**Comment by mheon** [comment]
We should cut an RC2 on Monday with this landed, I think. 1.6.0 final on Friday, maybe?

**Comment by rhatdan** [comment]
SGTM.  We'll have @edsantiago back as well.

**Pull request merged**